finalfrontier/src/sgd.rs

```diff
@@ -152,9 +151,17 @@ impl NegativeSamplingSGD {
         loss += self.negative_samples(model, input_embed, input_delta.view_mut(), output, lr);
 
         // Update the input embeddings with the accumulated gradient.
-        for &idx in input {
-            let input_embed = model.input_embedding_mut(idx as usize);
-            scaled_add(input_embed, input_delta.view(), 1.0);
+        match input {
+            Idx::Single(idx) => {
```

Comment:
Myabe this could be shortened if Idx was iterable? The both cases could be combined into one case that that iterates over an idx instance.